File: src/shadow_agent/permissions.py
"""Permission levels, command policy, and audit-friendly decisions."""

from __future__ import annotations

import re
from dataclasses import dataclass

from shadow_agent.config import PermissionLevel

READ_TOOLS = {
    "list_files",
    "read_file",
    "search_files",
    "search_text",
    "search_symbol",
    "git_status",
    "git_diff",
    "git_log",
    "git_branch",
    "update_plan",
    "update_todos",
}

WRITE_TOOLS = {
    "write_file",
    "edit_file",
    "create_directory",
    "move_file",
    "delete_file",
    "apply_patch",
    "exec",
    "kill",
    "git_add",
    "git_commit",
    "git_checkout",
}

ELEVATED_TOOLS = {
    "git_reset",
    "git_clean",
}

DANGEROUS_COMMAND = [
    re.compile(r"\bsudo\b"),
    re.compile(r"\brm\s+(-[a-zA-Z]*r[a-zA-Z]*f|-[a-zA-Z]*f[a-zA-Z]*r)\b"),
    re.compile(r"\bmkfs\b"),
    re.compile(r"\bdd\s+"),
    re.compile(r":\(\)\s*\{"),
    re.compile(r"\bchmod\s+(-R\s+)?777\b"),
    re.compile(r"(curl|wget).+\|\s*(ba)?sh"),
    re.compile(r"\b(shutdown|reboot|halt|poweroff)\b"),
    re.compile(r">\s*/dev/sd"),
    re.compile(r"\bgit\s+push\s+[^\n]*--force"),
    re.compile(r"\bgit\s+reset\s+--hard"),
    re.compile(r"\bgit\s+clean\s+-[a-zA-Z]*f"),
    re.compile(r"\bchown\s+-R\s+root\b"),
]

NETWORK_COMMAND = re.compile(r"\b(curl|wget|ssh|scp|rsync|nc|ncat|nmap|pip3?|uv|npm|pnpm|yarn)\b")
ROOT_COMMAND = re.compile(r"\bsudo\b|\b--privileged\b")


@dataclass
class PermissionDecision:
    allowed: bool
    reason: str
    needs_approval: bool = False
    elevated: bool = False


class PermissionGate:
    def __init__(
        self,
        level: PermissionLevel = PermissionLevel.WORKSPACE,
        *,
        require_approval_for_dangerous: bool = True,
        network: bool = False,
        allow_root: bool = False,
        auto_approve: bool = False,
    ) -> None:
        self.level = level
        self.require_approval_for_dangerous = require_approval_for_dangerous
        self.network = network
        self.allow_root = allow_root
        self.auto_approve = auto_approve

    def check_tool(self, tool_name: str) -> PermissionDecision:
        if tool_name in ELEVATED_TOOLS and self.level != PermissionLevel.ELEVATED:
            return PermissionDecision(False, f"{tool_name} requires elevated permissions")
        if tool_name in WRITE_TOOLS and self.level == PermissionLevel.READ_ONLY:
            return PermissionDecision(False, f"{tool_name} blocked in read-only mode")
        if tool_name in READ_TOOLS or tool_name in WRITE_TOOLS or tool_name in ELEVATED_TOOLS:
            return PermissionDecision(True, "ok")
        if self.level == PermissionLevel.READ_ONLY:
            return PermissionDecision(False, f"unknown tool {tool_name} blocked in read-only mode")
        return PermissionDecision(True, "ok")

    def check_command(self, command: str) -> PermissionDecision:
        if ROOT_COMMAND.search(command) and not self.allow_root:
            return PermissionDecision(False, "root/sudo is disabled")
        if any(pat.search(command) for pat in DANGEROUS_COMMAND):
            if self.level != PermissionLevel.ELEVATED:
                return PermissionDecision(
                    False,
                    "dangerous command requires elevated permissions",
                    needs_approval=True,
                    elevated=True,
                )
            if self.require_approval_for_dangerous and not self.auto_approve:
                return PermissionDecision(
                    False,
                    "dangerous command waiting for approval",
                    needs_approval=True,
                    elevated=True,
                )
        if not self.network and NETWORK_COMMAND.search(command):
            return PermissionDecision(False, "network commands are disabled")
        return PermissionDecision(True, "ok")

File: native/core/src/permissions.rs
//! Native tool policy. Two user-facing modes (`ask`, `allow_edits`) on top of
//! the advanced level (`read_only`, `workspace`, `elevated`):
//!
//! - web tools: allowed only when the task's web flag is on and the network
//!   mode is online; otherwise denied with the reason.
//! - file edits inside the project: `ask` asks with a readable summary,
//!   `allow_edits` allows. Paths outside the project are refused by the
//!   workspace layer after symlink resolution, before anything asks.
//! - shell, delete, Git history, background processes: always ask.
//! - destructive Git (reset/clean): denied below `elevated`, asked at `elevated`.
//! - privileged commands (sudo/su/pkexec/doas/run0): denied unless allow_root,
//!   and then still asked. They never run without a prompt.
//! - network-reaching shell commands: denied when shell network is off or the
//!   app is offline.
use crate::config::{Config, PermissionLevel, PermissionsConfig};
use serde_json::{json, Value};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Decision {
    Allow,
    Deny(String),
    Ask(String),
}

pub fn read_only(tool: &str) -> bool {
    matches!(
        tool,
        "system_info"
            | "list_files"
            | "read_file"
            | "search_files"
            | "search_text"
            | "search_symbol"
            | "workspace_symbols"
            | "goto_definition"
            | "find_references"
            | "get_diagnostics"
            | "get_type_signature"
            | "repo_map"
            | "search_code"
            | "mcp_sqlite_tables"
            | "mcp_sqlite_query"
            | "background_list"
            | "background_output"
            | "git_status"
            | "git_diff"
            | "git_log"
            | "update_plan"
            | "update_todos"
            | "web_fetch"
            | "web_search"
    )
}

pub fn web_tool(tool: &str) -> bool {
    matches!(tool, "web_fetch" | "web_search")
}

pub fn parallel_safe(tool: &str, args: &Value) -> bool {
    (read_only(tool) && !matches!(tool, "update_plan" | "update_todos"))
        || (tool == "git_branch" && args["create"].as_bool() != Some(true))
}

const PRIVILEGED: &[&str] = &["sudo", "doas", "su", "pkexec", "run0", "--privileged"];
const NETWORK: &[&str] = &[
    "curl", "wget", "ssh", "scp", "sftp", "rsync", "nc", "ncat", "netcat", "socat", "telnet",
    "ftp", "nmap", "pip", "pip3", "npm", "npx", "pnpm", "yarn", "uv", "uvx", "bunx", "pipx",
    "aria2c",
];
const DANGEROUS: &[&str] = &[
    "rm", "mkfs", "dd", "chmod", "chown", "reboot", "shutdown", "--force", "--hard",
];

fn words(command: &str) -> Vec<&str> {
    command
        .split(|c: char| !c.is_ascii_alphanumeric() && c != '_' && c != '-')
        .filter(|s| !s.is_empty())
        .collect()
}

/// Git commands that discard work or rewrite history. Lexical and
/// conservative: it can only add prompts, never remove them.
pub fn destructive_git(command: &str) -> bool {
    let words = words(command);
    let Some(start) = words.iter().position(|w| *w == "git") else {
        return false;
    };
    let rest = &words[start + 1..];
    let has = |w: &str| rest.contains(&w);
    has("reset")
        || has("clean")
        || has("rebase")
        || has("filter-branch")
        || has("filter-repo")
        || has("update-ref")
        || has("restore")
        || (has("checkout") && command.contains(" -- "))
        || (has("push")
            && (has("--force") || has("-f") || has("--force-with-lease") || has("--delete")))
        || (has("branch") && (has("-D") || has("--delete")))
        || (has("stash") && (has("drop") || has("clear")))
        || (has("reflog") && has("expire"))
        || (has("gc") && command.contains("--prune"))
}

/// A readable one-line description of a file edit, for the approval prompt.
pub fn edit_summary(tool: &str, args: &Value) -> String {
    let path = args["path"].as_str().unwrap_or("?");
    match tool {
        "write_file" => format!("Write {path}"),
        "edit_file" => format!("Edit {path}"),
        "create_directory" => format!("Create directory {path}"),
        "move_file" => format!(
            "Move {} to {}",
            args["src"].as_str().unwrap_or("?"),
            args["dest"].as_str().unwrap_or("?")
        ),
        "delete_file" => format!("Delete {path}"),
        "apply_patch" => {
            let patch = args["patch"]
                .as_str()
                .or_else(|| args["diff"].as_str())
                .unwrap_or("");
            let mut files: Vec<&str> = patch
                .lines()
                .filter_map(|line| {
                    line.strip_prefix("+++ ")
                        .map(|p| p.trim().trim_start_matches("b/"))
                        .filter(|p| *p != "/dev/null")
                        .or_else(|| line.strip_prefix("*** Update File: "))
                        .or_else(|| line.strip_prefix("*** Add File: "))
                        .or_else(|| line.strip_prefix("*** Delete File: "))
                        .map(str::trim)
                })
                .collect();
            files.dedup();
            if files.is_empty() {
                if let Some(path) = args["path"].as_str() {
                    files.push(path);
                }
            }
            let shown: Vec<_> = files.iter().take(6).copied().collect();
            let more = files.len().saturating_sub(shown.len());
            if shown.is_empty() {
                "Apply a patch".into()
            } else if more > 0 {
                format!("Apply a patch to {} and {more} more", shown.join(", "))
            } else {
                format!("Apply a patch to {}", shown.join(", "))
            }
        }
        _ => format!("Change files with {tool}"),
    }
}

pub fn check(config: &PermissionsConfig, tool: &str, args: &Value) -> Decision {
    use Decision::*;
    if web_tool(tool) {
        return if config.web {
            Allow
        } else if config.offline {
            Deny("The app is offline; web tools are unavailable".into())
        } else {
            Deny("Web tools are off for this task. Turn on web for the task (and keep the network mode online) to let the agent fetch pages.".into())
        };
    }
    if read_only(tool) || (tool == "git_branch" && args["create"].as_bool() != Some(true)) {
        return Allow;
    }
    if config.level == PermissionLevel::ReadOnly {
        return Deny(format!("{tool} is unavailable in read-only mode"));
    }
    match tool {
        "write_file" | "edit_file" | "apply_patch" | "create_directory" | "move_file" => {
            if config.approve_edits() {
                Ask(edit_summary(tool, args))
            } else {
                Allow
            }
        }
        "delete_file" => Ask(edit_summary(tool, args)),
        "git_add" => Ask("Stage files; repository clean filters may execute commands".into()),
        "git_commit" | "git_checkout" | "git_branch" => {
            Ask("Change repository history or the active branch".into())
        }
        "git_reset" | "git_clean" => {
            if config.level != PermissionLevel::Elevated {
                Deny("Destructive Git operations require elevated permissions".into())
            } else {
                Ask("Destructive Git operation: uncommitted work can be lost".into())
            }
        }
        "background_stop" => Ask("Stop a managed background process in this project".into()),
        "exec" | "background_start" => {
            let command = args["command"].as_str().unwrap_or("");
            let words = words(command);
            if words.iter().any(|word| PRIVILEGED.contains(word)) {
                if !config.allow_root {
                    return Deny(
                        "Root and privileged commands (sudo, su, pkexec, doas) are disabled".into(),
                    );
                }
                return Ask(
                    "Privileged command: it runs with elevated rights outside this project".into(),
                );
            }
            if words.iter().any(|word| NETWORK.contains(word)) {
                if config.offline {
                    return Deny("The app is offline; network commands are disabled".into());
                }
                if !config.network {
                    return Deny(
                        "Enable network commands in permissions before running this command".into(),
                    );
                }
            }
            if destructive_git(command) {
                return Ask(
                    "Destructive Git command: uncommitted work or history can be lost".into(),
                );
            }
            if config.shell_asks() {
                if tool == "background_start" {
                    return Ask("Start a project background process. It continues independently after this coding task, including cancellation, until stopped or the application closes. Its shell effects are not undone by rewind.".into());
                }
                return Ask("Run a shell command as your user; it can affect files and services beyond this project".into());
            }
            if config.require_approval_for_dangerous
                && words.iter().any(|word| DANGEROUS.contains(word))
            {
                Ask("Potentially destructive shell command".into())
            } else {
                Allow
            }
        }
        _ => Deny(format!("Unknown or unregistered tool: {tool}")),
    }
}

/// What ShadowCode can and cannot enforce for each vendor CLI, keyed by vendor
/// id. Returned with GET /api/config as `permissions.vendor_notes`.
pub fn vendor_notes(config: &Config) -> Value {
    let mode = match config.permissions.mode {
        crate::config::PermissionMode::Ask => "Ask before actions",
        crate::config::PermissionMode::AllowEdits => "Allow project edits",
    };
    json!({
        "native": format!("{mode}: enforced by ShadowCode for every tool call. Shell commands always ask; sudo/su/pkexec/doas are blocked; network commands follow the network setting; read-only tasks cannot change files."),
        "codex": "Codex runs in its own sandbox: workspace-write for tasks (read-only for Plan/Review) with its approval requests shown here. ShadowCode cannot enforce 'Ask before actions' for edits inside that sandbox; Codex decides which actions need approval.",
        "claude": "Claude Code permission prompts are routed to ShadowCode and you answer them here; Plan/Review uses Claude's plan mode. Claude's own settings files can pre-approve tools that ShadowCode never sees.",
        "cursor": "Cursor sends ACP permission requests that ShadowCode shows for approval; Plan/Review uses Cursor's plan mode when the agent offers it. Actions Cursor does not ask about are outside ShadowCode's control.",
        "grok": "Grok sends ACP permission requests that ShadowCode shows for approval. Grok advertises no read-only mode, so Plan/Review is not enforced by the runtime.",
        "antigravity": "Antigravity runs through Google's ACP agent server and asks ShadowCode before running commands or editing files, as Cursor and Grok do.",
        "network": "Network limits and the sudo block apply to ShadowCode's own tools only. Vendor CLIs use their own network access and sandbox."
    })
}
